src/core/internal/array.d

```diff
@@ -0,0 +1,147 @@
+module core.internal.array;
+import core.stdc.stdint : uintptr_t;
+
+/*
+ * This allows safe use of a pointer without dereferencing. Useful when you
+ * want to declare that all you care about is the pointer value itself. Keeps
+ * the pointer as a pointer to allow GC to properly track things.
+ *
+ * Note that PtrVal has all the same characteristics of a normal pointer,
+ * except it doesn't allow dereferencing, and doesn't allow pointer math
+ * without jumping into simple integer types.
+ */
+struct PtrVal(T)
+{
+    private T *_ptr;
+
+    /*
+     * Get the pointer as a non-dereferencable unsigned integer.
+     */
+    uintptr_t value() const @trusted
+    {
+        return cast(uintptr_t)_ptr;
+    }
+
+    /*
+     * Keep the mechanism that returns a signed pointer difference between pointers.
+     */
+    ptrdiff_t opBinary(string op : "-")(const(PtrVal) other) const @trusted
+    {
+        return _ptr - other._ptr;
+    }
+
+    // ditto
+    ptrdiff_t opBinary(string op : "-")(const(T*) other) const @trusted
+    {
+        return _ptr - other;
+    }
+
+    // ditto
+    ptrdiff_t opBinaryRight(string op : "-")(const(T*) other) const @trusted
+    {
+        return other - _ptr;
+    }
+
+    // disable subtraction between undefined types -- we don't want to
+    // accidentally subtract between the value() and some unknown type.
+    ptrdiff_t opBinary(string op : "-", X)(X) const @safe
+    {
+        static assert(0, "Cannot subtract between " ~ typeof(this).stringof ~ " and " ~ X.stringof);
+    }
+
+    /*
+     * Use this if you are in system code and wish to get back to unsafe
+     * pointer-land.
+     */
+    inout(T) *ptr() inout @system
```

Comment:
I think this shouldn't be `inout` since your returning a copy of the pointer and if the user wants a const pointer, then `PtrVal!(const T*)` should be used.